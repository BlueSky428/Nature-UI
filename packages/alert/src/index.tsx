import * as React from 'react';
import clx from 'clsx';
import { FiInfo, FiAlertCircle, FiAlertTriangle, FiCheckCircle } from 'react-icons/fi';

const SUBTLE_TEXT = 'text-gray-800';

export const ALERT_STATUSES = {
  error: {
    bg: 'bg-red-200',
    iconColor: 'text-red-600 mr-3',
    icon: FiAlertCircle,
    variant: {
      solid: 'bg-red-600 text-white',
      subtle: SUBTLE_TEXT,
    },
  },
  info: {
    bg: 'bg-blue-200',
    iconColor: 'text-blue-600 mr-3',
    icon: FiInfo,
    variant: {
      solid: 'bg-blue-600',
      subtle: SUBTLE_TEXT,
    },
  },
  success: {
    bg: 'bg-green-200',
    iconColor: 'text-green-600 mr-3',
    icon: FiCheckCircle,
    variant: {
      solid: 'bg-green-600 text-white',
      subtle: SUBTLE_TEXT,
    },
  },
  warning: {
    bg: 'bg-orange-200',
    iconColor: 'text-orange-600 mr-3',
    icon: FiAlertTriangle,
    variant: {
      solid: 'bg-orange-600 text-white',
      subtle: SUBTLE_TEXT,
    },
  },
};

interface AlertOptions {
  /**
   * The status of the alert
   */
  status?: keyof typeof ALERT_STATUSES;
  /**
   * The variant of the alert style to use
   */
  variant?: 'solid' | 'subtle';

  /**
   * Other tailwind utility classes or custom classnames you wish to include
   */
  className?: string;

  children?: React.ReactNode;
  /**
   * The component used for the root node.
   * Either a string to use a HTML element or a component.
   */
  component?: React.ElementType;
}

export type AlertProps = AlertOptions;

const BASE_STYLE = 'px-4 py-3 flex items-center';

const Alert: React.ForwardRefExoticComponent<AlertProps> = React.forwardRef(
  (props: AlertProps, ref: React.Ref<any>) => {
    const {
      className = '',
      status = 'success',
      children,
      variant = 'subtle',
      component: Component = 'div',
    } = props;

    // const Component = 'div';

    const VARIANT: string = ALERT_STATUSES[status].variant[variant];

    const componentClass = clx(BASE_STYLE, {
      [className]: className,
      [ALERT_STATUSES[status].bg]: status,
      [VARIANT]: variant,
    });

    const Icon = ALERT_STATUSES[status];
    const iconClasses = clx({
      [VARIANT]: variant,
      'mr-3': variant,
      [Icon.iconColor]: variant !== 'solid',
    });

    const IconComponent = Icon.icon;

    return (
      <Component className={componentClass} ref={ref}>
        {status && <IconComponent className={iconClasses} size={20} />}
        <span>{children ? children : 'This is an alert ðŸ™‚'}</span>
      </Component>
    );
  }
);

Alert.displayName = 'Alert';

export default Alert;
