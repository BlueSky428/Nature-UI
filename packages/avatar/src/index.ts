export * from './avatar';
export * from './avatarGroup';
