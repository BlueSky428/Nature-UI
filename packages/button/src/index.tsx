import * as React from 'react';
import clx from 'clsx';

interface ButtonProps {
  /**
   * The label of the button
   */
  'aria-label'?: string;
  /**
   * The text color of the button. Use a color key passed in theme.colors.
   */
  text?: string;
  /**
   * The background color of the button. Use a color key passed in theme.colors.
   */
  color?: string;
  /**
   * The variant of the button style to use.
   */
  variant?: 'outline' | 'ghost' | 'link' | 'solid' | 'none';
  /**
   * If true, the button will be disabled.
   */
  isDisabled?: boolean;
  /**
   * If true, the button will show a spinner.
   */
  isLoading?: boolean;
  /**
   * The label to show in the button when isLoading is true. If no text is passed, it only shows the spinner
   */
  loadingText?: string;
  size?: 'xs' | 'sm' | 'md' | 'lg' | 'none';
  /**
   * Other tailwind utility classes or custom classnames you wish to include
   */
  className?: string;
  /**
   * The lightness or how dark you want the background color to be
   */
  children?: React.ReactNode;
  /**
   * The component used for the root node.
   * Either a string to use a HTML element or a component.
   * defaults to "button"
   */
  component?: React.ElementType;
  id?: string;
  /**
   *  when the href prop is set it changes the tag name to a
   */
  href?: string;
  onClick?: () => any;
}

const darken = (value: string): string => {
  const splitStr = value.split('-');
  const dept = Number(splitStr[1]);
  const color = splitStr[0];

  if (dept >= 800) {
    return `${color}-${dept - 200}`;
  }

  return `${color}-${dept + 200}`;
};

const lighten = (value: string): string => {
  const color = value.split('-')[0];

  return `${color}-${100}`;
};

const Button = React.forwardRef((props: ButtonProps, ref: React.Ref<any>) => {
  const {
    component: Component = 'button',
    variant = 'solid',
    color = 'teal-500',
    text = 'white',
    size = 'md',
    children,
    ...rest
  } = props;
  const DEFAULT_CLASS = 'block focus:shadow-outline focus:outline-none rounded font-semibold flex';
  const STYLES = {
    solid: `bg-${color} text-${text} hover:bg-${darken(color)}`,
    outline: `bg-transparent hover:bg-${lighten(
      text
    )} text-${text} border border-${text} focus:border-transparent`,
    ghost: `hover:bg-${lighten(text)} text-${text}`,
    link: `hover:underline text-${text}`,
    md: 'px-4 py-2',
  };

  let BTNClass: string;

  if (variant === 'none') {
    BTNClass = '';
  } else {
    BTNClass = clx(DEFAULT_CLASS, {
      [STYLES[size]]: size && variant !== 'link',
      [STYLES[variant]]: variant,
    });
  }

  return (
    <Component className={BTNClass} ref={ref} {...rest}>
      {children ? children : 'Button'}
    </Component>
  );
});

export default Button;
