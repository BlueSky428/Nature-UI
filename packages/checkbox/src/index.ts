export * from './checkbox';
export * from './checkbox-group';
export * from './checkbox-types';
export * from './use-checkbox';
export * from './use-checkbox-group';
