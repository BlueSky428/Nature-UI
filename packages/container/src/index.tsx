import * as React from 'react';
import clx from 'clsx';

interface IContainer {
  children?: React.ReactNode;
  /**
   * Determine the max-width of the container.
   */
  size?: 'lg' | 'md' | 'sm' | 'xl' | 'xs';
  /**
   * The component used for the root node.
   * Either a string to use a HTML element or a component.
   */
  component?: React.ComponentType;
  className?: string;
}

const BASE_STYLE = 'mx-auto container';
const LG = 'px-20';
const MD = 'px-32';
const SM = 'px-48';
const XL = 'px-0';
const XS = 'px-64';

const Container: React.ForwardRefExoticComponent<IContainer> = React.forwardRef(
  (props: IContainer, ref: React.Ref<any>) => {
    const { component: Component = 'div', size = 'lg', children, className = '' } = props;

    const componentClass: string = clx(BASE_STYLE, {
      [LG]: size === 'lg',
      [MD]: size === 'md',
      [SM]: size === 'sm',
      [XL]: size === 'xl',
      [XS]: size === 'xs'
    });

    return (
      <Component className={componentClass} ref={ref}>
        <Component className={className}>{children}</Component>
      </Component>
    );
  }
);

Container.displayName = 'Container';

export default Container;
