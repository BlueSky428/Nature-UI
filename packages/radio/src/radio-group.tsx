import * as React from 'react';
import { PropsOf, forwardRef, nature, clsx } from '@nature-ui/system';
import { createContext, __DEV__ } from '@nature-ui/utils';

import {
  useRadioGroup,
  UseRadioGroupProps,
  UseRadioGroupReturn,
} from './use-radio-group';

export type RadioGroupContext = Pick<
  UseRadioGroupReturn,
  'onChange' | 'value' | 'name'
> & {
  size?: 'sm' | 'md' | 'lg' | number;
  color?: string;
};

const [
  RadioGroupContextProvider,
  useRadioGroupContext,
] = createContext<RadioGroupContext>({
  name: 'RadioGroupContext',
  strict: false,
});

export { useRadioGroupContext };

const DivTag = nature('div');

export type RadioGroupProps = UseRadioGroupProps &
  Omit<PropsOf<typeof DivTag>, 'onChange' | 'value' | 'defaultValue'> & {
    children: React.ReactNode;
    size?: 'sm' | 'md' | 'lg' | number;
  };

/**
 * Used for multiple radios which are bound in one group,
 * and it indicates which option is selected.
 *
 * @see Docs https://nature-ui.com/components/radio
 */
export const RadioGroup = forwardRef(
  (props: RadioGroupProps, ref: React.Ref<any>) => {
    const { size = 'md', children, className, color, ...hookProps } = props;

    const { value, onChange, getRootProps, name, htmlProps } = useRadioGroup(
      hookProps,
    );

    const group = React.useMemo(
      () => ({
        name,
        size,
        onChange,
        color,
        value,
      }),
      [size, name, onChange, value],
    );

    const groupProps = getRootProps({
      ref,
      ...htmlProps,
    });
    const _className = clsx('nature-radio-group', className);

    return (
      <RadioGroupContextProvider value={group}>
        <DivTag {...groupProps} className={_className}>
          {children}
        </DivTag>
      </RadioGroupContextProvider>
    );
  },
);

if (__DEV__) {
  RadioGroup.displayName = 'RadioGroup';
}
