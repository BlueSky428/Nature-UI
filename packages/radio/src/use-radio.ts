import { useFormControlContext } from '@nature-ui/form-control';
import { useBoolean, useControllableProp, useId } from '@nature-ui/hooks';
import { PropGetter } from '@nature-ui/react-utils';
import { ariaAttr, callAllHandler, dataAttr, Dict } from '@nature-ui/utils';
import { visuallyHiddenStyle } from '@nature-ui/visually-hidden';
import * as React from 'react';
import { InputDOMAttributes } from './../../react-utils/src/types';
import { useRadioGroupContext } from './radio-group';

/**
 * Prevent `onBlur` being fired when the checkbox label is touched
 */
const stop = (event: React.SyntheticEvent) => {
  event.preventDefault();
  event.stopPropagation();
};
export interface UseRadioProps {
  /**
   * id assigned to input
   */
  id?: string;
  /**
   * The name of the input field in a radio
   * (Useful for form submission).
   */
  name?: string;
  /**
   * The value to be used in the radio button.
   * This is the value that will be returned on form submission.
   */
  value?: string | number;
  /**
   * If `true`, the radio will be checked.
   * You'll need to pass `onChange` to update it's value (since it's now controlled)
   */
  isChecked?: boolean;
  /**
   * If `true`, the radio will be initially checked.
   */
  defaultChecked?: boolean;
  /**
   * If `true`, the radio will be disabled
   */
  isDisabled?: boolean;
  /**
   * If `true` and `isDisabled` is true, the radio will remain
   * focusable but not interactive.
   */
  isFocusable?: boolean;
  /**
   * If `true`, the radio will be read-only
   */
  isReadOnly?: boolean;
  /**
   * If `true`, the radio button will be invalid. This sets `aria-invalid` to `true`.
   */
  isInvalid?: boolean;
  /**
   * If `true`, the radio button will be invalid. This sets `aria-invalid` to `true`.
   */
  isRequired?: boolean;
  /**
   * Function called when checked state of the `input` changes
   */
  onChange?: (event: React.ChangeEvent<HTMLInputElement>) => void;
  'data-radiogroup'?: any;
  /**
   * Refers to the `id` of the element that labels the radio element.
   */
  'aria-describedby'?: string;
}

export interface RadioState {
  isInvalid: boolean | undefined;
  isFocused: boolean;
  isChecked: boolean;
  isActive: boolean;
  isHovered: boolean;
  isDisabled: boolean | undefined;
  isReadOnly: boolean | undefined;
  isRequired: boolean | undefined;
}

export const useRadio = (props: UseRadioProps = {}) => {
  const {
    defaultChecked,
    isChecked: isCheckedProp,
    isFocusable,
    isDisabled: isDisabledProp,
    isReadOnly: isReadOnlyProp,
    isRequired: isRequiredProp,
    onChange,
    isInvalid: isInvalidProp,
    name,
    value,
    id: idProp,
    'data-radiogroup': dataRadioGroup,
    'aria-describedby': ariaDescribedBy,
    ...htmlProps
  } = props;

  const uuid = useId(undefined, 'radio');

  const formControl = useFormControlContext();
  const group = useRadioGroupContext();

  const isWithinRadioGroup = !!group || !!dataRadioGroup;
  const isWithinFormControl = !!formControl;

  let id = isWithinFormControl && !isWithinRadioGroup ? formControl.id : uuid;
  id = idProp ?? id;

  const isDisabled = isDisabledProp ?? formControl?.isDisabled;
  const isReadOnly = isReadOnlyProp ?? formControl?.isReadOnly;
  const isRequired = isRequiredProp ?? formControl?.isRequired;
  const isInvalid = isInvalidProp ?? formControl?.isInvalid;
  const { onFocus, onBlur } = formControl ?? {};

  const [isFocused, setFocused] = useBoolean();
  const [isHovered, setHovering] = useBoolean();
  const [isActive, setActive] = useBoolean();

  const [isCheckedState, setChecked] = React.useState(Boolean(defaultChecked));

  const [isControlled, isChecked] = useControllableProp(
    isCheckedProp,
    isCheckedState,
  );

  const handleChange = React.useCallback(
    (event: React.ChangeEvent<HTMLInputElement>) => {
      if (isReadOnly || isDisabled) {
        event.preventDefault();

        return;
      }

      if (!isControlled) {
        setChecked(event.target.checked);
      }

      onChange?.(event);
    },
    [isControlled, isDisabled, isReadOnly, onChange],
  );

  const onKeyDown = React.useCallback(
    (event: KeyboardEvent) => {
      if (event.key === ' ') {
        setActive.on();
      }
    },
    [setActive],
  );

  const onKeyUp = React.useCallback(
    (event: KeyboardEvent) => {
      if (event.key === ' ') {
        setActive.off();
      }
    },
    [setActive],
  );

  const getRadioProps: PropGetter = React.useCallback(
    (props = {}, ref = null) => ({
      ...props,
      ref,
      'data-active': dataAttr(isActive),
      'data-hover': dataAttr(isHovered),
      'data-disabled': dataAttr(isDisabled),
      'data-invalid': dataAttr(isInvalid),
      'data-checked': dataAttr(isChecked),
      'data-focus': dataAttr(isFocused),
      'data-readonly': dataAttr(isReadOnly),
      'aria-hidden': true,
      onMouseDown: callAllHandler(props.onMouseDown, setActive.on),
      onMouseUp: callAllHandler(props.onMouseUp, setActive.off),
      onMouseEnter: callAllHandler(props.onMouseEnter, setHovering.on),
      onMouseLeave: callAllHandler(props.onMouseLeave, setHovering.off),
    }),
    [
      isActive,
      isHovered,
      isDisabled,
      isInvalid,
      isChecked,
      isFocused,
      isReadOnly,
      setActive.on,
      setActive.off,
      setHovering.on,
      setHovering.off,
    ],
  );

  const getInputProps: PropGetter<InputDOMAttributes, InputDOMAttributes> =
    React.useCallback(
      (props: Dict = {}, ref = null) => {
        const trulyDisabled = isDisabled && !isFocusable;
        return {
          ...props,
          ref,
          type: 'radio',
          name,
          value,
          id,
          onChange: callAllHandler(props.onChange, handleChange),
          onBlur: callAllHandler(props.onBlur, setFocused.off),
          onFocus: callAllHandler(props.onFocus, setFocused.on),
          onKeyDown: callAllHandler(props.onKeyDown, onKeyDown),
          onKeyUp: callAllHandler(props.onKeyUp, onKeyUp),
          checked: isChecked,
          disabled: trulyDisabled,
          readOnly: isReadOnly,
          required: isRequired,
          'aria-invalid': ariaAttr(isInvalid),
          'aria-required': ariaAttr(isRequired),
          'aria-disabled': ariaAttr(trulyDisabled),
          'data-readonly': dataAttr(isReadOnly),
          'aria-describedby': ariaDescribedBy,
          style: visuallyHiddenStyle,
        };
      },
      [
        isDisabled,
        isFocusable,
        id,
        name,
        value,
        handleChange,
        onBlur,
        setFocused,
        onFocus,
        onKeyDown,
        onKeyUp,
        isChecked,
        isReadOnly,
        isRequired,
        isInvalid,
        ariaDescribedBy,
      ],
    );

  const getLabelProps: PropGetter = (props: Dict = {}, ref = null) => {
    return {
      ...props,
      ref,
      style: {
        ...props.style,
        touchAction: 'none',
      },
      onMouseDown: callAllHandler(props.onMouseDown, stop),
      onTouchStart: callAllHandler(props.onTouchState, stop),
      'data-disabled': dataAttr(isDisabled),
      'data-checked': dataAttr(isChecked),
      'data-invalid': dataAttr(isInvalid),
    };
  };

  const getRootProps: PropGetter = (props, ref = null) => ({
    ...props,
    ref,
    'data-disabled': dataAttr(isDisabled),
    'data-checked': dataAttr(isChecked),
    'data-invalid': dataAttr(isInvalid),
  });

  const state: RadioState = {
    isInvalid,
    isFocused,
    isChecked,
    isActive,
    isHovered,
    isDisabled,
    isReadOnly,
    isRequired,
  };

  return {
    state,
    getCheckboxProps: getRadioProps,
    getInputProps,
    getLabelProps,
    getRootProps,
    htmlProps,
  };
};

export type UseRadioReturn = ReturnType<typeof useRadio>;
