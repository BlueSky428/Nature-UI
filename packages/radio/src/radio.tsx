import { nature, clsx, PropsOf, css } from '@nature-ui/system';
import { darken, __DEV__ } from '@nature-ui/utils';
import * as React from 'react';

import { useRadioGroupContext } from './radio-group';
import { useRadio, UseRadioProps } from './use-radio';

const DivTag = nature('div');

type Omitted = 'onChange' | 'defaultChecked' | 'checked';

type BaseControlProps = Omit<PropsOf<typeof DivTag>, Omitted>;

export type RadioProps = UseRadioProps &
  BaseControlProps & {
    /**
     * The spacing between the checkbox and it's label text
     * @default 0.5rem
     */
    spacing?: string;
    /**
     * If `true`, the radio will occupy the full width of it's parent container
     */
    isFullWidth?: boolean;
    size?: 'sm' | 'md' | 'lg' | number;
    colorScheme?: string;
    wrapperClass?: string;
  };

const StyledWrapper = nature('label');

const _SIZES = {
  sm: '0.75rem',
  md: '1rem',
  lg: '1.25rem',
};

/**
 * Radio
 *
 * Radio component is used in forms when a user needs to select a single value from
 * several options.
 *
 * @see Docs https://nature-ui.com/components/radio
 */
export const Radio = React.forwardRef(
  (props: RadioProps, ref: React.Ref<any>) => {
    const group = useRadioGroupContext();

    const {
      spacing = '0.5rem',
      size = group?.size || 'md',
      children,
      isFullWidth,
      colorScheme = 'blue-500',
      wrapperClass = '',
      ...radioProps
    } = props;

    let { isChecked } = props;

    if (group?.value && props.value) {
      isChecked = group.value === props.value;
    }

    let { onChange } = props;

    if (props.value && group?.onChange) {
      onChange = group.onChange;
    }

    const name = group?.name || props?.name;
    const _size = typeof size === 'string' ? _SIZES[size] : `${size}px`;

    const {
      getInputProps,
      getCheckboxProps,
      getLabelProps,
      htmlProps: rest,
    } = useRadio({
      ...radioProps,
      isChecked,
      onChange,
      name,
    });

    const { style, ...inputProps } = getInputProps({ ref });
    const { ...checkboxProps } = getCheckboxProps(rest);

    const _css = css`
      &::before {
        content: '';
        display: inline-block;
        position: relative;
        width: 50%;
        height: 50%;
        border-radius: 50%;
        background: currentColor;
      }
    `;

    const dark = darken(colorScheme, 100);

    const _focus = typeof checkboxProps['data-focus'] !== 'undefined';
    const _checked = typeof checkboxProps['data-checked'] !== 'undefined';
    const _invalid = typeof checkboxProps['data-invalid'] !== 'undefined';
    const _hover = typeof checkboxProps['data-hover'] !== 'undefined';
    const _disabled = typeof checkboxProps['data-disabled'] !== 'undefined';

    const _className = clsx(
      props.className,
      'nature-radio__control inline-flex items-center justify-center flex-shrink-0 border-2 border-solid rounded-full text-white transition-all duration-150',
      _css,
      {
        [`bg-${colorScheme} border-${colorScheme}`]: _checked && !_invalid,
        ring: _focus,
        'border-red-600': _invalid,
        [`bg-${dark} border-${dark}`]: _hover && _checked && !_invalid,
        'bg-gray-300 text-gray-300': _disabled,
        'text-gray-500': _disabled && _checked,
      },
    );

    return (
      <StyledWrapper
        className={clsx(
          wrapperClass,
          'nature-radio inline-flex items-center align-top',
          {
            'opacity-50': _disabled,
          },
        )}
        css={{
          width: isFullWidth ? 'full' : undefined,
        }}
      >
        <input
          className='nature-radio__input'
          style={style as any}
          {...inputProps}
        />
        <DivTag
          css={{
            width: _size,
            height: _size,
          }}
          {...checkboxProps}
          className={_className}
        />
        {children && (
          <DivTag
            className='nature-radio__label select-none'
            {...getLabelProps()}
            css={{
              marginLeft: spacing,
            }}
          >
            {children}
          </DivTag>
        )}
      </StyledWrapper>
    );
  },
);

if (__DEV__) {
  Radio.displayName = 'Radio';
}
