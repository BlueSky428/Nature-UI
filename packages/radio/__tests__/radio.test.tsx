import * as React from 'react';
import { render } from '@nature-ui/test-utils';

import { Radio, useRadio, UseRadioProps } from '../src';

describe('@nature-ui/radio', () => {
  test('Radio renders correctly', () => {
    const { asFragment } = render(<Radio size='lg' color='blue-500' />);

    expect(asFragment()).toMatchSnapshot();
  });

  test('has proper aria and data attributes', async () => {
    const Component = (props: UseRadioProps = {}) => {
      const { getCheckboxProps, getInputProps } = useRadio(props);

      return (
        <label>
          <input data-testid='input' {...getInputProps()} />
          <div data-testid='checkbox' {...getCheckboxProps()} />
        </label>
      );
    };
    const utils = render(<Component name='name' value='' id='id' />);
    const input = utils.getByTestId('input');
    const checkbox = utils.getByTestId('checkbox');

    expect(input).toHaveAttribute('name', 'name');
    expect(input).toHaveAttribute('id', 'id');
    expect(input).toHaveAttribute('value', '');
    expect(input).not.toBeDisabled();
    expect(input).not.toHaveAttribute('aria-required');
    expect(input).not.toHaveAttribute('aria-invalid');
    expect(input).not.toHaveAttribute('aria-disabled');
    expect(checkbox).toHaveAttribute('aria-hidden', 'true');
    expect(checkbox).not.toHaveAttribute('data-active');
    expect(checkbox).not.toHaveAttribute('data-hover');
    expect(checkbox).not.toHaveAttribute('data-checked');
    expect(checkbox).not.toHaveAttribute('data-focus');
    expect(checkbox).not.toHaveAttribute('data-readonly');

    // render with various flags enabled
    utils.rerender(<Component isDisabled isInvalid isReadOnly isRequired />);

    expect(input).toHaveAttribute('aria-required');
    expect(input).toHaveAttribute('aria-invalid');
    expect(input).toHaveAttribute('aria-disabled');
    expect(input).toBeDisabled();
    expect(checkbox).toHaveAttribute('data-readonly');

    // input is not truly disabled if focusable
    utils.rerender(<Component isDisabled isFocusable />);
    expect(input).not.toBeDisabled();
  });
});
