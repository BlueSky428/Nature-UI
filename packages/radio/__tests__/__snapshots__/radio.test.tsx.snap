// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@nature-ui/radio Radio renders correctly 1`] = `
<DocumentFragment>
  <label
    class="nature-radio inline-flex items-center align-top css-1y18hzf-Radio"
  >
    <input
      class="nature-radio__input"
      style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
      type="radio"
      value=""
    />
    <div
      aria-hidden="true"
      class="nature-radio__control inline-flex items-center justify-center flex-shrink-0 border-2 border-solid rounded-full text-white transition-all duration-150 css-kklk9c css-1wncg9q-Radio"
    />
  </label>
</DocumentFragment>
`;
